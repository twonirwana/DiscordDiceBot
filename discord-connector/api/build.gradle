plugins {
    id 'java'
}

group 'de.janno'
version '1.0-SNAPSHOT'

repositories {
    mavenCentral()
}

dependencies {
    implementation "io.projectreactor:reactor-core:3.4.17"
    implementation 'com.google.guava:guava:31.1-jre'

    compileOnly 'org.projectlombok:lombok:1.18.24'
    annotationProcessor 'org.projectlombok:lombok:1.18.22'

    testImplementation 'org.junit.jupiter:junit-jupiter-api:5.8.2'
    testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine:5.8.2'
}

java {
    toolchain {
        languageVersion.set(JavaLanguageVersion.of(14))
    }
}

test {
    useJUnitPlatform()
}