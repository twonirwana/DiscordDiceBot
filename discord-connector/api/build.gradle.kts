plugins {
    id("java")
}

repositories {
    mavenCentral()
}

dependencies {
    implementation(libs.reactor.core)
    implementation(libs.guava)
    implementation(libs.commons.lang3)

    compileOnly(libs.lombok)
    annotationProcessor(libs.lombok)

    testImplementation(libs.junit.jupiter.api)
    testRuntimeOnly(libs.junit.jupiter.engine)
    testImplementation(libs.assertj.core)
}

java {
    toolchain {
        languageVersion.set(JavaLanguageVersion.of(21))
    }
}

tasks.getByName<Test>("test") {
    useJUnitPlatform()
}