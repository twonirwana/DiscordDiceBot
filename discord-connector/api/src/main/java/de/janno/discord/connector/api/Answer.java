package de.janno.discord.connector.api;

import com.google.common.base.Strings;
import lombok.NonNull;
import lombok.Value;
import reactor.util.annotation.Nullable;

import java.util.List;
import java.util.stream.Collectors;

@Value
public class Answer {
    private static final String MINUS = "\u2212";
    @NonNull
    String title;
    @Nullable
    String content;
    @NonNull
    List<Field> fields;

    public String toShortString() {
        List<String> fieldStringList = fields.stream()
                .map(Field::toShortString)
                .collect(Collectors.toList());

        return String.format("%s %s %s", title,
                        Strings.isNullOrEmpty(content) ? "" : content,
                        fieldStringList.isEmpty() ? "" : fieldStringList.toString())
                .replace("▢", "0")
                .replace("＋", "+")
                .replace(MINUS, "-")
                .replace("*", "");
    }

    @Value
    public static class Field {
        @NonNull
        String name;
        @NonNull
        String value;
        boolean inline;

        public String toShortString() {
            return String.format("%s %s", name, value);
        }
    }
}
