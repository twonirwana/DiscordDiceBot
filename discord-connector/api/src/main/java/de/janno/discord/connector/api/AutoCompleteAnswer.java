package de.janno.discord.connector.api;

import lombok.NonNull;
import lombok.Value;
import org.apache.commons.lang3.StringUtils;

@Value
public class AutoCompleteAnswer {
    @NonNull
    String name;
    @NonNull
    String value;

    public AutoCompleteAnswer(String name, String value) {
        this.name = StringUtils.abbreviate(name, 100);
        this.value = StringUtils.abbreviate(value, 100);
    }
}
