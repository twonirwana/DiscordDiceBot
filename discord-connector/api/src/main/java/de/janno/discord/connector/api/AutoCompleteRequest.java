package de.janno.discord.connector.api;

import lombok.Value;

import java.util.List;

@Value
public class AutoCompleteRequest {
    String focusedOptionName;
    String focusedOptionValue;
    List<OptionValue> optionValues;

}
