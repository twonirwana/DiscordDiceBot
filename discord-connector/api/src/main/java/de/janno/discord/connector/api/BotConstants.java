package de.janno.discord.connector.api;

public final class BotConstants {
    public static final String LEGACY_DELIMITER_V2 = "\u0000";
    public static final String CUSTOM_ID_DELIMITER = "\u001e";
    public static final String LEGACY_DELIMITER_V1 = ",";
    public static final String LEGACY_CONFIG_SPLIT_DELIMITER_REGEX = String.format("[%s%s]", LEGACY_DELIMITER_V2, LEGACY_DELIMITER_V1);

}
