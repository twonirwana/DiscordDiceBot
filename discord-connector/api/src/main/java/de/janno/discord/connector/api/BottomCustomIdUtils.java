package de.janno.discord.connector.api;

import com.google.common.base.Preconditions;
import lombok.NonNull;
import org.apache.commons.lang3.StringUtils;

public final class BottomCustomIdUtils {
    public static final String CUSTOM_ID_DELIMITER = "\u001e";
    private static final String LEGACY_DELIMITER_V2 = "\u0000";
    private static final String LEGACY_DELIMITER_V1 = ",";
    public static final String LEGACY_CONFIG_SPLIT_DELIMITER_REGEX = String.format("[%s%s]", LEGACY_DELIMITER_V2, LEGACY_DELIMITER_V1);
    private static final int COMMAND_NAME_INDEX = 0;
    private static final int BUTTON_VALUE_INDEX = 1;

    static public @NonNull String createButtonCustomId(@NonNull String commandId, @NonNull String buttonValue) {
        return commandId + CUSTOM_ID_DELIMITER + buttonValue;
    }

    public static boolean isLegacyCustomId(@NonNull String customId) {
        return !customId.contains(CUSTOM_ID_DELIMITER);
    }

    public static @NonNull String getButtonValueFromCustomId(@NonNull String customId) {
        Preconditions.checkArgument(StringUtils.countMatches(customId, CUSTOM_ID_DELIMITER) == 1, "'%s' contains not the correct number of delimiter", customId);
        return customId.split(CUSTOM_ID_DELIMITER)[BUTTON_VALUE_INDEX];
    }

    public static @NonNull String getCommandNameFromCustomIdWithPersistence(@NonNull String customId) {
        Preconditions.checkArgument(StringUtils.countMatches(customId, CUSTOM_ID_DELIMITER) == 1, "'%s' contains not the correct number of delimiter", customId);
        return customId.split(CUSTOM_ID_DELIMITER)[COMMAND_NAME_INDEX];
    }

    /**
     * will be removed when almost all users have switched to the persisted button id
     */
    public static boolean matchesLegacyCustomId(@NonNull String customId, @NonNull String commandId) {
        return customId.matches("^" + commandId + LEGACY_CONFIG_SPLIT_DELIMITER_REGEX + ".*");
    }

}
