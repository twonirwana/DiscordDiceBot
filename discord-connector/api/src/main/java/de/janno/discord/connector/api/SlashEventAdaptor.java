package de.janno.discord.connector.api;

import de.janno.discord.connector.api.message.EmbedDefinition;
import de.janno.discord.connector.api.message.MessageDefinition;
import de.janno.discord.connector.api.slash.CommandInteractionOption;
import lombok.NonNull;
import reactor.core.publisher.Mono;

import java.util.Optional;

public interface SlashEventAdaptor extends DiscordAdapter {
    Optional<String> checkPermissions();

    Optional<CommandInteractionOption> getOption(@NonNull String actionStart);

    Mono<Void> reply(@NonNull String message);

    Mono<Void> replyEmbed(@NonNull EmbedDefinition embedDefinition, boolean ephemeral);

    Mono<Long> createButtonMessage(@NonNull MessageDefinition messageDefinition);

    long getChannelId();

    String getCommandString();

    Mono<Requester> getRequester();

    Mono<Void> createResultMessageWithEventReference(EmbedDefinition answer);

    boolean isValidAnswerChannel(long channelId);

}
