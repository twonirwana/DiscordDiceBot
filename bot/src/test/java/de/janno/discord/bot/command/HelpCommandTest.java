package de.janno.discord.bot.command;

import de.janno.discord.connector.api.SlashEventAdaptor;
import org.junit.jupiter.api.Test;
import reactor.core.publisher.Mono;
import reactor.test.StepVerifier;

import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

class HelpCommandTest {

    HelpCommand underTest = new HelpCommand();

    @Test
    void getCommandId() {
        assertThat(underTest.getCommandId()).isEqualTo("help");
    }

    @Test
    void handleSlashCommandEvent() {
        SlashEventAdaptor slashEventAdaptor = mock(SlashEventAdaptor.class);
        when(slashEventAdaptor.replyEmbed(any(), anyBoolean())).thenReturn(Mono.empty());

        Mono<Void> res = underTest.handleSlashCommandEvent(slashEventAdaptor, () -> UUID.fromString("00000000-0000-0000-0000-000000000000"));
        StepVerifier.create(res).verifyComplete();
    }
}