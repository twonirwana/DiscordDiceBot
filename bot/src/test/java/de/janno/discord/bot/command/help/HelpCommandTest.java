package de.janno.discord.bot.command.help;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import de.janno.discord.connector.api.SlashEventAdaptor;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import reactor.core.publisher.Mono;
import reactor.test.StepVerifier;

import java.util.Locale;
import java.util.UUID;

import static org.mockito.Mockito.*;

@ExtendWith(SnapshotExtension.class)
class HelpCommandTest {

    HelpCommand underTest = new HelpCommand();

    private Expect expect;

    @Test
    public void getCommandDefinition() {
        expect.toMatchSnapshot(underTest.getCommandDefinition());
    }

    @Test
    public void getId() {
        expect.toMatchSnapshot(underTest.getCommandId());
    }

    @Test
    void handleSlashCommandEvent() {
        SlashEventAdaptor slashEventAdaptor = mock(SlashEventAdaptor.class);
        when(slashEventAdaptor.replyWithEmbedOrMessageDefinition(any(), anyBoolean())).thenReturn(Mono.empty());

        Mono<Void> res = underTest.handleSlashCommandEvent(slashEventAdaptor, () -> UUID.fromString("00000000-0000-0000-0000-000000000000"), Locale.ENGLISH);
        StepVerifier.create(res).verifyComplete();
    }
}