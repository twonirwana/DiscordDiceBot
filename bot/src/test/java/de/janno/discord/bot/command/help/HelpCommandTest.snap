de.janno.discord.bot.command.help.HelpCommandTest.getCommandDefinition=[
  {
    "name" : "help",
    "description" : "Help to the commands and links for further information",
    "nameLocales" : [ ],
    "descriptionLocales" : [
      {
        "locale" : "de",
        "description" : "Hilfe zu Befehle und Links zu weiterer Information"
      },
      {
        "locale" : "pt_BR",
        "description" : "Ajuda para os comandos e links para e mais informações"
      }
    ],
    "options" : [ ]
  }
]


de.janno.discord.bot.command.help.HelpCommandTest.getId=[
  "help"
]